ENT-12414: Use `librsync`'s Stream API in `GET <FILENAME>` requests

larsewi opened the pull request:
- **Added file stream API**
- **New network protocol version v4 - filestream**

**TODO:**
 - [ ] Depends on https://github.com/cfengine/buildscripts/pull/1520
 - [x] Depends on https://github.com/NorthernTechHQ/libntech/pull/230
 - [x] Check that errors in different places in the code does not make client/server hang
 - [x] Add `librsync` as dependency in CI
 - [ ] Should be merged before hand https://gitlab.com/Northern.tech/CFEngine/jenkins-vms/-/merge_requests/59

[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=11510)](https://ci.cfengine.com/job/pr-pipeline/11510/)

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+            HEADER_SIZE);
+        return false;
+    }
+    memcpy(&header, recv_buffer, HEADER_SIZE);
```
You could use `&header` as the buffer and avoid this `memcpy()`.

larsewi replied:
Yup, I also thought so. But `TLSRecv()` expects a buffer of size `CF_BUFSIZE` and that you never request more than `CF_BUFSIZE - 1` Bytes. And the reason why, is that it actually writes a terminating `\0` at the end of the message. I spent a long time trying to figure out why my program crashed with memory violations :stuck_out_tongue_winking_eye: There is a comment some lines above. But it probably deserves more explanation.

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+    *eof = header & (1 << 1);
+
+    /* Extract message length */
+    *len = header >> 4;
```
This requires `len` (`size_t`) to be at least 3 bytes, right? Probably that's guaranteed, but I'd add a static assert for this to avoid potential nasty surprises.

larsewi replied:
If it was left shift, then yes, 3 bytes. However, with right shift, two bytes should do. The following:
```C
    assert(sizeof(*len) >= 2); /* "It's probably guaranteed, but let's make
                                * sure to avoid potentially nasty surprises."
                                *  - @vpodzime */
```
still would not hurt.

vpodzime replied:
Right, it's 12 bits, not 16. I got confused by `uint16_t`.

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+    /* Read payload */
+    if (*len > 0)
+    {
+        ret = TLSRecv(conn, recv_buffer, *len);
```
I take it `TLSRecv()` does some retries to avoid short reads/recvs? Worth a comment.

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+ * @param ... The format string arguments
+ * @return true on success, otherwise false
+ */
+static bool SendError(SSL *conn, bool flush, const char *fmt, ...)
```
There's some `#pragma` to mark this as a printf-like function and let the compiler do all sorts of checks. Please add it.

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+    bufs.next_out = out_buf;
+    bufs.avail_out = MESSAGE_SIZE; /* We cannot send more using the protocol */
+
+    do
```
Any chance we could de-duplicate what's in this `do-while` with the one in `SendDelta()`? It's very long anyway...

larsewi replied:
Yes, I think it should be possible. The log messages would have to be more generic and the code would probably be a bit harder to understand. I will create a follow-up ticket to investigate this further.

larsewi replied:
https://northerntech.atlassian.net/browse/CFE-4456

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+            GetErrorStr());
+
+        /* At this point the server will not be expecting any more messages
+         * from the client as far as the File Stream API is conserned. Hence,
```
*concerned*

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -592,6 +592,31 @@ bool FileStreamServe(SSL *conn, const char *filename)
+ * @note -1 on error is quite handy, because rs_sig_args() iterprets it as
+ *       unknown file size
+ */
+static rs_long_t GetFileSize(FILE *file)
```
Do we need to worry about symlinks in this function?

vpodzime replied:
Yes. That's why I wanted to suggest `lstat()`, but then I realized it was probably not the desired thing to do. And it made me think how the new protocol handles symlinks. It just copies the contents of the symlink target into a file on the other side, right? Worth a comment.

larsewi replied:
Yup! Added a note in the doc strings :wink:

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -592,6 +592,31 @@ bool FileStreamServe(SSL *conn, const char *filename)
+ *
+ * @param file The file pointer
+ * @return the file size or -1 on error
+ * @note -1 on error is quite handy, because rs_sig_args() iterprets it as
```
*interprets*

vpodzime commented on libcfnet/file_stream.c:
```diff
@@ -592,6 +592,31 @@ bool FileStreamServe(SSL *conn, const char *filename)
+        return -1;
+    }
+
+    return S_ISREG(sb.st_mode) ? sb.st_size : -1;
```
`(rs_long_t)` in both cases.

vpodzime reviewed (commented):
Looks good to me otherwise. Very nice! Thanks for taking the hard road! :clap: :beers: I'm very excited about this!

larsewi commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+
+    /* Send header */
+    header = htons(header);
+    int ret = TLSSend(conn, (char *) &header, HEADER_SIZE);
```
The main motivation for having this macro was that I would not have to cast the result to an `int`

larsewi commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,943 @@
+ * @return true on success, otherwise false
+ */
+static bool __SendMessage(
+    SSL *conn, const char *msg, size_t len, bool eof, bool err)
```
Changing that argument to an `int` would make me have to do a lot of casting where ever these functions are called. The fact that it is an unsigned integer along with the `assert(len <= MESSAGE_SIZE)` also makes sure any values out of range are caught.

vpodzime replied:
Shall we change `TLSSend()` to take a `size_t` value instead? That makes most sense, of course.

larsewi replied:
Yes, I can create a follow-up ticket for this :wink:

larsewi replied:
Then we can also use `sizeof(header)` without having to cast it to an int

larsewi replied:
https://northerntech.atlassian.net/browse/CFE-4455

vpodzime reviewed (approved):
Let's see if it works! :grin: :crossed_fingers: :beers:

craigcomstock commented on configure.ac:
```diff
@@ -521,6 +521,20 @@ CF3_WITH_LIBRARY(pcre2, [
 
+dnl librsync
+
+AC_ARG_WITH([librsync], [AS_HELP_STRING([--with-librsync[[=PATH]]], [Spefify librsync path])], [], [with_librsync=yes])
```
```suggestion
AC_ARG_WITH([librsync], [AS_HELP_STRING([--with-librsync[[=PATH]]], [Specify librsync path])], [], [with_librsync=yes])
```

craigcomstock commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,1013 @@
+            {
+                /* The job requires more data, but the input buffer is full */
+                Log(LOG_LEVEL_ERR,
+                    "Insufficient buffer capacity to compute delta");
```
again, maybe include avail_in and MESSAGE_SIZE

craigcomstock commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,1013 @@
+    assert(basis != NULL);
+    assert(dest != NULL);
+
+    /* The input buffer has to be twice the message size, so that if can fit a
```
```suggestion
    /* The input buffer has to be twice the message size, so that it can fit a
```

craigcomstock commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,1013 @@
+                /* The job requires more data, but we cannot fit another
+                 * message into the input buffer */
+                Log(LOG_LEVEL_ERR,
+                    "Insufficient buffer capacity to receive file stream delta");
```
add avail_in and MESSAGE_SIZE to error message?

craigcomstock commented on libcfnet/file_stream.h:
```diff
@@ -0,0 +1,104 @@
+/**
+ * @brief Reply with unspecified server refusal
+ *
+ * E.g., use this function to when the resource does not exist or access is
```
```suggestion
 * E.g., use this function when the resource does not exist or access is
```

craigcomstock reviewed (commented):
Some comments and little typos.

vpodzime reviewed (approved):
ACK

larsewi commented on libcfnet/file_stream.c:
```diff
@@ -0,0 +1,1013 @@
+    /* librsync has rs_file_size() which basically does the same thing.
+     * However, it is only available in the most recent versions of the
+     * library. Hence, we provide our own implementation in order to have
+     * some backwards compatibility in terms of librsync versions. */
```
Added some more info in the comment. When it comes to minimum version checks, it does not seem like we have  this for any of the other dependencies. If you think we should have these checks, then we should probably make a separate ticket for this.

craigcomstock replied:
I think what we have is fine. You have a workaround for older versions and we don't have to go down a rabbit hole of trying to figure out where our code-base with the workaround would stop working as we get into older versions of the library.

larsewi commented:
Just tested this manually on Windows and it works like a charm.
```
PS C:\Program Files\Cfengine\bin> .\cf-agent.exe -KI
    info: Copied file '/var/cfengine/masterfiles/promises.cf' from '172.31.39.27' to 'C:\promises.cf.cfnew'
    info: Moved 'C:\promises.cf.cfnew' to 'C:\promises.cf'
    info: Updated file 'C:\promises.cf' from '172.31.39.27:/var/cfengine/masterfiles/promises.cf'
```
Did however find a separate (surprisingly unrelated) issue in `cf-net.exe` (see [ENT-12511](https://northerntech.atlassian.net/browse/ENT-12511))
